README_developers.md

```diff
@@ -91,128 +85,356 @@ $ # For a fix
+overloading instead. For example:
 ```cpp
-// Bad, prefer polymorphism
+// No
 void print_double(double val);
 void print_str(std::string_view val);
 
-// Good
+// Yes
 void print(double val);
 void print(std::string_view val);
+```
+
+## Declarations
+
+### Variables
+
+Have each variable declaration have its own type declaration
+For example:
 
-// Good
-class FooBar
+```cpp
+// Yes
+int a;
+int b;
+int c;
+
+// No
+int a, b, c;
+```
+
+### Long declaration lines
+When breaking methods declaration into multiple lines:
+- have a limit of ~125 characters per line
+- keep the first argument on the same line with the method name
+- align arguments on the subsequent lines with the first argument
+- if class name or return type long, place the function name at the beginning of the line and put the return type and class name qualification on the previous line (see example below)
+
+```c++
+// Yes: all arguments on the same line
+void ClassName::some_method(double beta, double alpha) const
 {
-private:
-  double* data_device_;
-public:
-  void to_device();
 }
 
-// Bad
-class Foo_bar {
-  private:
-    double* dataDevice;
-  public:
-    void toDevice();
+// Yes: each argument on a separate line
+void ClassName::some_method(double beta,
+                            hiopMatrix& W,
+                            double alpha,
+                            const hiopMatrix& X) const
+{
+}
+
+// Yes: when return type + class name + method name + first argument would overrun the 125 characters line limit, have method name on the next line
+long_return_type SomeVeryLongClassNameIMeanLong::
+some_method(double beta,
+            hiopMatrix& W,
+            double alpha,
+            const hiopMatrix& X) const
+{
+}
+
+// No: multiple arguments per line when using more than one line for the declaration 
+void ClassName::some_method(double beta, hiopMatrix& W,
+                            double alpha, const hiopMatrix& X) const
+{
+}
+
+// No: arguments not aligned behind the opening parenthesis
+void ClassName::some_method(
+  double beta,
+  hiopMatrix& W,
+  double alpha,
+  const hiopMatrix& X) 
+{
 }
 ```
 
-## Comments
 
+## Comments
 - Don't comment what can be expressed in code
 - Comment intent, not implementation
 - Block comments (`/* comment */`) are preferred for comments three lines or longer
+
+### Documenting methods
+
+Each method should be documented using Doxygen style comments. At minimum,
+there should be description of method arguments, template parameters (if
+applicable), return value, pre- and post-conditions. Use template as below.
+Input and output arguments should be clearly marked as such (e.g. `@param[in]`
+for inputs).
+
+```c++
+/**
+ * @brief <BRIEF DESCRIPTION>
+ *
+ * @param <NAME> <DESCRIPTION>
+ * @tparam <NAME> <DESCRIPTION>
+ * @return <DESCRIPTION OF RETURN VALUE>
+ *
+ * @pre <PRECONDITION>
+ * @post <POSTCONDITION>
+ *
+ * LONGER DESCRIPTION, RUNTIME, EXAMPLES, ETC 
+ */
+```
```

Comment:
I would perhaps just add that detailed commens should be next to the implementation rather than in the header file. If you make changes in the implementation you are less likely to forget to update comments if they are right there.

Reply:
I might suggest pre and post conditions remain next to the interface - that seems important for the users to know about, no?

Reply:
Do we need to add brief descriptions in the header file?
For inherited classes, do we need to add the same description to all the inherited functions?

Reply:
> I might suggest pre and post conditions remain next to the interface - that seems important for the users to know about, no?

I agree

Reply:
> Do we need to add brief descriptions in the header file?
> For inherited classes, do we need to add the same description to all the inherited functions?

I think that would be good practice.

Reply:
I've updated the doc with the above.